### ISSUE TITLE
Unify the nixos and nixpkgs channel structures?

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#7659]
Currently, as far as I can tell, the `nixos` channel is just the `nixpkgs` channel with a level of indirection and a tiny wrapper to put `nixpkgs` inside a `pkgs` attribute. I've seen a few people on IRC get confused by the weird distinction and it screwed me up a bit when I was trying to publish my own channels too.

Is there a deep reason for the distinction? I realize that lots of people expect it now, but it seems like simplifying it would be good in the long run. Even if we unified it, we could keep a self-referential `pkgs` attribute inside the top-level of `nixpkgs` so that existing references to `nixos.pkgs` wouldn't break. The attribute could emit a deprecation warning when forced.

Am I missing something?

**Comment by Ericson2314** [comment]
So those two channels have nothing today with packages vs services, nixos vs nix? Then absolutely! Case in point, they certainly confused me.

**Comment by vcunat** [comment]
You can do ``nix-env -iA nixpkgs.pkgs.foo`` (I just tried). And the first ``nixpkgs`` bit comes only from the name **you** give locally to the channel. On one machine, for example, I have ``p`` instead:
```
$ nix-channel --list
p http://nixos.org/channels/nixpkgs-unstable
```

There's a difference independent of this issue: different jobs must succeed in order for a channel to be updated by Hydra.

**Comment by edolstra** [comment]
The motivation for having separate NixOS and Nixpkgs channels is:
* The NixOS channels only include Linux packages. We don't want NixOS updates to be blocked by (say) broken Darwin packages.
* Conversely, the Nixpkgs channel doesn't depend on NixOS tests, so Darwin users (and non-NixOS Linux users) still can get updates if the tests are failing.

**Comment by copumpkin** [comment]
@edolstra hmm okay, I think I buy the separation of the channels, but I'm still not sure why they need to have distinct structures. Why `nixos.pkgs` vs `nixpkgs`? It seems like There could be two (or three) distinct channels server-side (with different update criteria) that exist locally with the same name and structure, which would then let us say easily that "nix-env -iA pkgs.git` will work regardless of your system. I'm really just trying to avoid "branching" in our documentation/intuition.

**Comment by edolstra** [comment]
The separate structure is mostly a remnant of the time when Nixpkgs and NixOS were in separate repositories. It would be great to get rid of the distinction.

**Comment by copumpkin** [comment]
Awesome! Does this sound like a reasonable plan forward to you?

```nix
let 
  annoyance = "Yo, quit using nixos.pkgs; we're unifying the nixpkgs and nixos structures. See https://github.com/NixOS/nixpkgs/issues/7659 for more information.";
in rec {
  nixpkgs = {
    pkgs = builtins.trace annoyance nixpkgs;
    # all the usual nixpkgs stuff
  };
  nixos = nixpkgs;
}
```

(`nixpkgs` and `nixos` are obviously channels instead of nix values, but I wanted a compact representation of the idea)

**Comment by vcunat** [comment]
As I mentioned above, nixos and nixpkgs are just names that **you choose locally** for the channels.

**Comment by copumpkin** [comment]
I realize that. What am I saying that indicates I misunderstand that?

**Comment by copumpkin** [comment]
@vcunat to expand on my last message, the point is that the `nixos.org/channels/nixos-unstable` and `nixos.org/channels/nixos-14.12`, etc. serve up `nixexprs.tar.bz2` files that contain a different top-level expression than the `nixos.org/channels/nixpkgs-unstable` channel does. The `nixos-*` channels have a `pkgs` attribute that contains roughly what the `nixpkgs-unstable` channel contains.

By convention, people add local named channels called `nixos` pointing at the `nixos-*` server-side channels, and then type things like `nix-env -iA nixos.pkgs.git` to get at the equivalent of `nixpkgs.git`. I'm proposing giving the `nixexprs.tar.bz2` in the `nixos-*` channels the same exact contents as the one in the `nixpkgs-unstable` channel. To support the old convention of referring to `nixos.pkgs.git`, I would add `pkgs` attribute to the top-level `nixpkgs` that refers back to itself with a warning message.

I _think_ this will result in a fairly smooth transition, along with warnings to move away from the `nixos.pkgs` convention. Am I missing something?

**Comment by vcunat** [comment]
``nix-env -iA ``channel-name``.pkgs.glib`` does work already, regardless of being on a nixos or nixpkgs channel. That channel-name string is what you choose, which is good, because some people might want to have multiple channels at once.

Side note: that ``pkgs`` attribute within nixpkgs is used for ``packageOverrides`` magic, so I'm not exactly sure that ``nixpkgs.pkgs

**Comment by vcunat** [comment]
Maybe I just misunderstand your exact intentions.

**Comment by copumpkin** [comment]
Oh, I didn't realize that `pkgs` was already in `nixpkgs`. I wouldn't be changing its meaning, but if it's used internally I could see adding the `trace` message being annoying. Urgh :frowning:

**Comment by cstrahan** [comment]
@copumpkin: I _think_ that the `pkgs` attr is akin to `self` in e.g. the haskellngPackages sense; I've seen plenty of examples of this in people's `~/.nixpkgs/config.nix` (annotated with my understanding):

``` nix
{ pkgs }: { # pkgs is the fully configured packages, after overrides
  packageOverrides = super: let self = super.pkgs; in rec { # super is without overrides, self is with
    # ... overrides go here ...
  }
}
```

**Comment by Ericson2314** [comment]
@vcunat @cstrahan I don't see any `haskellngPackages.self` in nix repl, so I assume the `packageOverrides` idiom can be made to work without making self/pkgs an attribute in the attribute set being overridden? Obviously we want overriding to work, but if it can be done without exposing `pkgs` all the better, as at it confusing and redundant (IMO) to expose self/pkgs an attribute.

**Comment by philandstuff** [comment]
There's currently a big difference between nixos-unstable and nixpkgs-unstable: about 10 days. nixos-unstable last updated 16 May, nixpkgs-unstable last updated 26 May.

The latest nixos-unstable is unbootable for me due to #7859.  Is it possible to use nixpkgs-unstable to upgrade or would that be dumb?  I've tried searching for documentation which explains the differences between the channels but I've been unable to find any.  I also don't even see entries for nixos-unstable or nixpkgs-unstable in hydra.nixos.org (which is where I assume the channels get built) so I'm not sure when to expect the channels to be updated.

This is all very confusing for someone quite new to nix and nixos.

**Comment by vcunat** [comment]
@philandstuff: it is. The difference is in criteria of updating (tests etc.) and in set of binaries uploaded to binary cache. Otherwise, it's just like checking out a different nixpkgs version in git. There's also nixos-unstable-small.

**Comment by Ericson2314** [comment]
One nice thing about haskell-ng is you can repeatedly override the package set. Nixpkgs, on the other hand, defines a override function `applyGlobalOverrides` only temporary for integrating the config. Now I'm even more strongly in the camp that we should add a `overrideNg` that works exactly like haskell-ng, and deprecate `nixpkgs.pkgs`.

**Comment by Ericson2314** [comment]
https://github.com/unisonweb/platform/blob/topic/nix2/env.nix#L22-L28 This is an example for what I have to do instead because the override I described above doesn't exist.

**Comment by Ericson2314** [comment]
Starting to work on "modernizing" the overrides system to address these issues. https://github.com/Ericson2314/nixpkgs/commit/af5f0e61e00262ac848e58ba29195342940749f4 diverges and I don't know why -- probably missing something obvious.

@peti Tagging you because you presumably wrote the haskell-ng overrides system and that is what I am (eventually) trying to emulate.

**Comment by Ericson2314** [comment]
Here is that same divergence problem narrowed down a bit https://github.com/Ericson2314/nixpkgs/commit/a11db14020f96a73a9a1d91f48c4294a7cfdd92b . I think it could be a real issue that `pkgsOrig` is fixed with itself, in that that means that overrides may not be depended upon by other pkgs in all cases.

**Comment by peti** [comment]
I haven't really followed this thread in detail, so I may be way off here, but one source of trouble I see is that the top-level package set is recursive, i.e. it's defined as `rec { ... }`. The Haskell NG package set makes that recursion explicit through the `self` argument, i.e. it's defined as `self: { ... }`. Mixing those two idioms might not work as expected. (I'd have to spend more time thinking about this to come up with a concrete example, though.)

**Comment by vcunat** [comment]
@peti: it seems natural to have the references recursive by default, as that's what people mostly mean IMHO (and what all-packages.nix currently does anyway), but you can still allow a non-recursive reference on-demand (``super`` you called it, IIRC?).

**Comment by Ericson2314** [comment]
@peti Hmm? I see don't see any top-level `rec { ... }` in all-packages.nix. The tighter fixpoint is almost that however, and I think the current double fix point has basically the same issues as mixing the `rec { ... }` with `self: { ... }`. My commit comes from https://github.com/Ericson2314/nixpkgs/tree/override which slowly starts to override.

@vcunat As I understand it, basic model is this:
```nix
let
  # vanilla functions should not "see" themselves or overrides will not effect dependencies
  # If one wants to override without effecting dependencies they can just use `//`.
  # Giving them `super` too would also allow them to observe if a package was overridden.
  defaults  = self: { ... };
 
  # in accordance with above, only "first level" is vanilla follow the dependency graph
  super = defaults self;
  
  # overrider just uses super when operation does not have fixpoint, e.g.
  # `myPkg = myPkg // { buildInputs = super.myPkg.buildInputs ++ [ optionalDep ] };`
  self = defaults self // overrider self super;
in self
```
What the haskell-ng stuff does (IIUC) is allow one to do:
```nix
(/*..*/((vanillaPkgs
  .override { overrider = self: super: { ... }; })
  .override { overrider = self: super: { ... }; })
# ..
  .override { overrider = self: super: { ... }; })
```
Which is pretty magical :). But the final value is exactly what you would get if you defined `overrider` in my first code listing as the composition of all the `overrider`s in my second code listing.

@everybody

My hope would be to have all-packages just contain the big defaults function, a second file do the fix point magic a la haskell-ng, and a third file add the config from ~/.config or wherever as the "first override" as that is a rather impure operation and it would be nice to have access an unconfigured nixpkgs.

**Comment by Ericson2314** [comment]
For an example in the wild, https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/agda/default.nix has something pretty similar with regards to the `extension` function.

**Comment by copumpkin** [comment]
Beautiful, thanks!!
